projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,154 @@
+  // runs the users callback if not NULL
+  if (storage->channel_callback[current_channel] != NULL) {
+    storage->channel_callback[current_channel](storage->current_channel,
+                                               storage->callback_context[current_channel]);
+  }
+  // Update so that the ADS1015 reads from the next channel
+  do {
+    current_channel = (current_channel + 1) % NUM_ADS1015_CHANNELS;
+  } while (!storage->channel_enable[current_channel]);
+  prv_set_channel(storage, current_channel);
+}
+
+// initiates ads1015 by setting up registers and enabling ALRT/RDY Pin
+StatusCode ads1015_init(Ads1015Storage *storage, I2CPort i2c_port, Ads1015Address i2c_addr,
+                        GPIOAddress *ready_pin) {
+  if ((storage == NULL) || (ready_pin == NULL)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  storage->i2c_port = i2c_port;
+  storage->i2c_addr = i2c_addr;
+  storage->ready_pin = *ready_pin;
+  for (uint8_t channel = 0; channel < NUM_ADS1015_CHANNELS; channel++) {
+    storage->channel_enable[channel] = false;
+    storage->channel_readings[channel] = 0;
+    storage->channel_callback[channel] = NULL;
+    storage->callback_context[channel] = NULL;
+  }
+
+  status_ok_or_return(prv_reset(storage));
+  status_ok_or_return(prv_setup_register(
+      storage, ADS1015_ADDRESS_POINTER_CONFIG, CONFIG_REGISTER_MSB(ADS1015_CHANNEL_0),
+      CONFIG_REGISTER_LSB));  // sets the starting channel to channel 0
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_LO_THRESH,
+                                         ADS1015_LO_THRESH_REGISTER_MSB,
+                                         ADS1015_LO_THRESH_REGISTER_LSB));
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_HI_THRESH,
+                                         ADS1015_HI_THRESH_REGISTER_MSB,
+                                         ADS1015_HI_THRESH_REGISTER_LSB));
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,  //
+  };
+  InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  status_ok_or_return(gpio_init_pin(&storage->ready_pin, &gpio_settings));
+  return gpio_it_register_interrupt(&storage->ready_pin, &it_settings, INTERRUPT_EDGE_RISING,
+                                    prv_interrupt_handler, storage);
+}
+
+// This function enable/disables channels, and registers callbacks for each channel.
+// It also registers the interrupt handler on ALRT/RDY pin.
+StatusCode ads1015_configure_channel(Ads1015Storage *storage, Ads1015Channel channel, bool enable,
+                                     Ads1015Callback callback, void *context) {
+  if (storage == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  // if no channels were already enabled, set to given channel
+  if (!storage->channel_enable[0] && !storage->channel_enable[1] && !storage->channel_enable[2] &&
```

Comment:
I feel like this is not a very clean way of doing this. I would suggest using a bitset or a loop.

Reply:
I took this out. It is not necessary to do it if I set the channel for the first time in the init function.